cpp/src/arrow/types/primitive.h

```diff
@@ -54,30 +54,54 @@ class PrimitiveArray : public Array {
-  NAME(int32_t length, const std::shared_ptr<Buffer>& data,     \
-      int32_t null_count = 0,                                   \
-      const std::shared_ptr<Buffer>& null_bitmap = nullptr) :   \
-      PrimitiveArray(std::make_shared<TypeClass>(), length,     \
-          data, null_count, null_bitmap) {}                     \
-                                                                \
-  bool EqualsExact(const NAME& other) const {                   \
-    return PrimitiveArray::EqualsExact(                         \
-        *static_cast<const PrimitiveArray*>(&other));           \
-  }                                                             \
-                                                                \
-  const T* raw_data() const {                                   \
-    return reinterpret_cast<const T*>(raw_data_);               \
-  }                                                             \
-                                                                \
-  T Value(int i) const {                                        \
-    return raw_data()[i];                                       \
-  }                                                             \
+#define NUMERIC_ARRAY_DECL(NAME, TypeClass, T)                              \
+class NAME : public PrimitiveArray {                                        \
+ public:                                                                    \
+  using value_type = T;                                                     \
+  using PrimitiveArray::PrimitiveArray;                                     \
+                                                                            \
+  NAME(int32_t length, const std::shared_ptr<Buffer>& data,                 \
+      int32_t null_count = 0,                                               \
+      const std::shared_ptr<Buffer>& null_bitmap = nullptr) :               \
+      PrimitiveArray(std::make_shared<TypeClass>(), length,                 \
+          data, null_count, null_bitmap) {}                                 \
+                                                                            \
+  bool EqualsExact(const NAME& other) const {                               \
+    return PrimitiveArray::EqualsExact(                                     \
+        *static_cast<const PrimitiveArray*>(&other));                       \
+  }                                                                         \
+                                                                            \
+  std::shared_ptr<NAME> slice(int32_t start) const{                         \
+    return slice(start, length_);                                           \
+  }                                                                         \
+                                                                            \
+  std::shared_ptr<NAME> slice(int32_t start, int32_t length) const{         \
```

Comment:
Yes, thats what i thought at first.
But, thare are two reasons that I put them here.
1.such as Int32Array etc., they are defined by '# define'.
2.If I put them in PrimitiveArray class, these methods should need more arguments.
   Also i don't think PrimitiveArray need slice method.
Finally, I will attempt to define a template slice method in PrimitiveArray, or using other
approaches to rewrite them.
Thanks again. : )